### PR TITLE
Implement caffe export

### PR DESCRIPTION
**Pull request by okuta**
Please check #3572.

### CODE REVIEW

**Comment by mitmul** [comment]
@okuta Is it still WIP?

**Comment by okuta** [comment]
I changed title.
Cloud you start review this?

File: chainer/caffe_export.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+                with open(self.caffemodel + ".txt", 'w') as f:
+                    f.write(google.protobuf.text_format.MessageToString(net))
+
+
+def caffe_export(input, output, directory=None,
+                 export_params=True, graph_name='Graph'):
+    """Export computational graph as Caffe format.
+
+    Args:
+        input (~chainer.Variable):
```
Comment (mitmul) [inline chainer/caffe_export.py]: Could you explain what the `input` should be here?

File: chainer/caffe_export.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        output (~chainer.Variable):
```
Comment (mitmul) [inline chainer/caffe_export.py]: Could you explain what the `output` should be here?

File: chainer/caffe_export.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        directory (str): The directory used for saving the resulting Caffe
+            model. If None, nothing is saved to the disk.
+        export_params (bool): If True, this function exports all the parameters
+            included in the given model at the same time. If False, the
+            exported ONNX model doesn't include any parameter values.
```
Comment (mitmul) [inline chainer/caffe_export.py]: I think this function doesn't export ONNX model. Please fix this explanation.

File: chainer/caffe_export.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+                f.write(net.SerializeToString())
+            if self.debug:
+                import google.protobuf.text_format
+                with open(self.caffemodel + ".txt", 'w') as f:
+                    f.write(google.protobuf.text_format.MessageToString(net))
+
+
+def caffe_export(input, output, directory=None,
+                 export_params=True, graph_name='Graph'):
+    """Export computational graph as Caffe format.
```
Comment (mitmul) [inline chainer/caffe_export.py]: Could you add an example to use this function here? And please also add the list of supported functions to be exported and some comments of what users should be noted if there's any restrictions or limitations. And it's also better to describe what architectures are tested.

**Comment by okuta** [comment]
I fixed.

File: chainer/exporters/caffe.py [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+            if self.debug:
+                import google.protobuf.text_format
+                with open(self.caffemodel + ".txt", 'w') as f:
+                    f.write(google.protobuf.text_format.MessageToString(net))
+
+
+def export(model, args, directory=None,
+           export_params=True, graph_name='Graph'):
+    """Export computational graph as Caffe format.
+
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: As I stated before, please write down the list of supported functions that can be included in the given model and some comments of what users should be noted if there's any restrictions or limitations here. And it's also better to describe what architectures are tested.

File: chainer/exporters/caffe.py [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+from chainer.links.caffe.protobuf3 import caffe_pb2 as caffe_pb
+from chainer import variable
+
+
+_function_types = (function.Function, function_node.FunctionNode)
+
+
+def _add_blob(layer, shape, data):
+    # The following part is ridiculously slow!!
+    # TODO(who): Replace with C++ extension call
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: Who should `who` be?

File: chainer/exporters/caffe.py [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+
+    def _get_layer_name(self, layer):
+        """Generate layer name like "Convolution2DFunction-10-2".
+
+        The first number means rank of the layer (depth from the top),
+        and the second number is for preventing duplication
+        (different layer objects can have same rank)
+
+        :param layer: Function object
+        :return: string unique to functions
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: The docstring format is different from others (this docstring is not exposed to users, though).

**Comment by mitmul** [comment]
A flake8 error is occurring. Please check the test results.

File: chainer/exporters/caffe.py [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+
+
+def export(model, args, directory=None,
+           export_params=True, graph_name='Graph'):
+    """Export computational graph as Caffe format.
+
+    Args:
+        model (~chainer.Chain): The model object you want to export in ONNX
+            format. It should have :meth:`__call__` method because the second
+            argment ``args`` is directly given to the model by the ``[]``
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: This: \`\`[]\`\` should be \`\`()\`\`

File: chainer/exporters/caffe.py [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+                f.write(net.SerializeToString())
+            if self.debug:
+                import google.protobuf.text_format
+                with open(self.caffemodel + ".txt", 'w') as f:
+                    f.write(google.protobuf.text_format.MessageToString(net))
+
+
+def export(model, args, directory=None,
+           export_params=True, graph_name='Graph'):
+    """Export computational graph as Caffe format.
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: Export "a" computational graph (Please insert "a")

File: chainer/exporters/caffe.py [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+       >>> caffe.export(Model(), [self.x], None, True, 'test')
+
+    """
+
+    assert isinstance(args, (tuple, list))
+    if len(args) != 1:
+        raise NotImplementedError()
+    for i in args:
+        assert isinstance(i, variable.Variable)
+    with function.force_backprop_mode(), chainer.using_config('train', False):
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: This configuration changes should be documented.

**Comment by mitmul** [comment]
And please import `from chainer import utils` and call `utils.experimental('chainer.exporters.caffe.export')` at the top of `export()` function.

**Comment by okuta** [comment]
I fixed.

File: chainer/exporters/caffe.py [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+        - :func:`~chainer.functions.reshape`
+        - :func:`~chainer.functions.add`
+
+        This function can export at least following networks.
+
+        - GoogLeNet
+        - ResNet
+        - VGG
+
+        And, this function use testing (evaluation) mode.
```
Comment (mitmul) [inline chainer/exporters/caffe.py]: It might not be enough for users. Is there any intention to omit the explanation about it forces backprop mode?

Comment (okuta) [reply chainer/exporters/caffe.py]: I think using force backprop mode is not important information for users.

Comment (mitmul) [reply chainer/exporters/caffe.py]: I see. Now I think this docstring looks fine.

**Comment by okuta** [comment]
Thanks! I will add.

**Comment by mitmul** [comment]
@okuta Thanks for adding the README.md. Could you check the Jenkins test result? The doc test is failing due to this WARNING:
```
WARNING: [autosummary] failed to import 'chainer.exporters.caffe.export': no module named chainer.exporters.caffe.export
```

**Comment by okuta** [comment]
@mitmul  The test was passed.

**Comment by mitmul** [comment]
LGTM

**Pull request merged**